sdk/core/System.ClientModel/src/Message/MultipartContent.cs

```diff
@@ -0,0 +1,295 @@
+
+        public MultipartContentSubpart(BinaryContent content, Dictionary<string, string> headers)
+        {
+            Content = content;
+            Headers = headers;
+        }
+
+        /// <summary>
+        /// Writes the <see cref="Headers"/> and <see cref="Content"/> of this subpart to the provided <see cref="Stream"/>.
+        /// </summary>
+        /// <param name="stream">The <see cref="Stream"/> to write to.</param>
+        /// <param name="cancellationToken">The <see cref="CancellationToken"/> to signal cancellation of the write operation.</param>
+        public void WriteTo(Stream stream, CancellationToken cancellationToken)
+        {
+            // Write headers on a new line
+            stream.Write(_crLfBytes, 0, _crLfBytes.Length);
+
+            // Write the headers to stream
+            foreach (var header in Headers)
+            {
+                string headerString = $"{header.Key}{ColonSpace}{header.Value}{CrLf}";
+                byte[] headerBytes = Encoding.UTF8.GetBytes(headerString);
+                stream.Write(headerBytes, 0, headerBytes.Length);
+            }
+
+            // Add another line
+            stream.Write(_crLfBytes, 0, _crLfBytes.Length);
+
+            // Write the content to stream
+            Content.WriteTo(stream, cancellationToken);
+        }
+
+        /// <summary>
+        /// Writes the <see cref="Headers"/> and <see cref="Content"/> of this subpart to the provided <see cref="Stream"/>.
+        /// </summary>
+        /// <param name="stream">The <see cref="Stream"/> to write to.</param>
+        /// <param name="cancellationToken">The <see cref="CancellationToken"/> to signal cancellation of the write operation.</param>
+        public async Task WriteToAsync(Stream stream, CancellationToken cancellationToken)
+        {
+            // Write headers on a new line
+            await stream.WriteAsync(_crLfBytes, 0, _crLfBytes.Length, cancellationToken).ConfigureAwait(false);
+
+            // Write the headers to stream
+            foreach (var header in Headers)
+            {
+                string headerString = $"{header.Key}{ColonSpace}{header.Value}{CrLf}";
+                byte[] headerBytes = Encoding.UTF8.GetBytes(headerString);
+                await stream.WriteAsync(headerBytes, 0, headerBytes.Length, cancellationToken).ConfigureAwait(false);
+            }
+
+            // Add another line
+            await stream.WriteAsync(_crLfBytes, 0, _crLfBytes.Length, cancellationToken).ConfigureAwait(false);
+
+            // Write the content to stream
+            await Content.WriteToAsync(stream, cancellationToken).ConfigureAwait(false);
+        }
+    }
+
+    #endregion
+
+    #region BinaryContent return helper
+    /// <summary>
+    /// A helper class to return a <see cref="MultipartContent"/> as a <see cref="BinaryContent"/>.
+    /// </summary>
+    private class MultipartBinaryContent : BinaryContent
```

Comment:
Non-blocking question: I'm curious why we decided not to inherit MultipartContent from BinaryContent.  Was there a reason?

Reply:
I don't recall there being a clear consensus either way in the meeting. The main benefit of not having the inheritance is that the `BinaryContent.Create` methods won't show up in intellisense.
When I was writing this, I personally thought not having the inheritance made the API a bit cleaner. I also liked the consistency of having `.ToBinaryContent()` and `ToRequestContent()` in both scenarios vs only in Azure.Core. But I don't have a strong preference. One benefit of having the inheritance would be that creating nested multipart requests is easier. The current implementation requires something like this:

```csharp
var innerContent = new MultipartContent();
var outerContent = new MultipartContent(innerContent.ToBinaryContent());
```

Reply:
I ended up changing this because the name of the type needed to change to avoid name conflicts with the existing MultipartContent type. The clearest alternative is to use MultipartBinaryContent and inherit.